flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java

```diff
@@ -0,0 +1,248 @@
+					return o -> new LongWritable((Long) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaFloatObjectInspector) {
+				if (((JavaFloatObjectInspector) inspector).preferWritable()) {
+					return o -> new FloatWritable((Float) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			} else if (inspector instanceof JavaDoubleObjectInspector) {
+				if (((JavaDoubleObjectInspector) inspector).preferWritable()) {
+					return o -> new DoubleWritable((Double) o);
+				} else {
+					return IdentityConversion.INSTANCE;
+				}
+			}
+		}
+
+		throw new FlinkHiveUDFException(
+			String.format("Flink doesn't support convert object conversion for %s yet", inspector));
+	}
+
+	/**
+	 * Unwraps an object from Hive object to Flink object with an ObjectInspector.
+	 */
+	public static Object unwrap(ObjectInspector inspector, Object data) {
+		if (data == null) {
+			return null;
+		}
+
+		if (inspector instanceof VoidObjectInspector) {
+			return null;
+		}
+
+		if (inspector instanceof PrimitiveObjectInspector) {
+			if (inspector instanceof BooleanObjectInspector) {
+				BooleanObjectInspector oi = (BooleanObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveJavaObject(data);
+			} else if (inspector instanceof StringObjectInspector) {
+				StringObjectInspector oi = (StringObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.getPrimitiveWritableObject(data).toString() :
+					oi.getPrimitiveJavaObject(data);
+			} else if (inspector instanceof ByteObjectInspector) {
+				ByteObjectInspector oi = (ByteObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveJavaObject(data);
+			} else if (inspector instanceof ShortObjectInspector) {
+				ShortObjectInspector oi = (ShortObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveWritableObject(data);
+			} else if (inspector instanceof IntObjectInspector) {
+				IntObjectInspector oi = (IntObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveWritableObject(data);
+			} else if (inspector instanceof LongObjectInspector) {
+				LongObjectInspector oi = (LongObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveWritableObject(data);
+			} else if (inspector instanceof FloatObjectInspector) {
+				FloatObjectInspector oi = (FloatObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveWritableObject(data);
+			} else if (inspector instanceof DoubleObjectInspector) {
+				DoubleObjectInspector oi = (DoubleObjectInspector) inspector;
+
+				return oi.preferWritable() ?
+					oi.get(data) :
+					oi.getPrimitiveWritableObject(data);
+			}
+		}
+
+		// TODO: handle complex types like struct, list, and map
```

Comment:
May not only complex types, for example, spark handle:
```
    case _: WritableStringObjectInspector => StringType
    case _: JavaStringObjectInspector => StringType
    case _: WritableHiveVarcharObjectInspector => StringType
    case _: JavaHiveVarcharObjectInspector => StringType
    case _: WritableHiveCharObjectInspector => StringType
    case _: JavaHiveCharObjectInspector => StringType
```

Reply:
IIUC, the example you give is in `inspectorToDataType()` and it's used to get a Spark data type from a Hive inspector, and shouldn't that match our `toFlinkType()` rather than our `unwrap()`?

Reply:
```
      case _: JavaHiveVarcharObjectInspector =>
        withNullSafe { o =>
            val s = o.asInstanceOf[UTF8String].toString
            new HiveVarchar(s, s.length)
        }
      case _: JavaHiveCharObjectInspector =>
        withNullSafe { o =>
            val s = o.asInstanceOf[UTF8String].toString
            new HiveChar(s, s.length)
          }
```
wrap also have these.

Reply:
I see what you mean now. Primitive types like char, varchar, timestamp, date, decimal, etc will added immediately afterwards